### DISCUSSION TITLE
Who ordered that?

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#10087] [Using MicroPython]
I often think of Isidor Rabi's quip "Who ordered that?" in reference to the unexpected appearance of the muon in the 1930s. It's how I feel when dealing with python types some days, actually most days.

Needed to check the first byte of a file to validate that it was a .mpy file.  mpy files start with an M  (eg b'M\x05.....). Since I've given up trying to second guess python I sought detail 
```
>>> mpy=b'M\x05'; print(mpy[0], type(mpy[0]))
77 <class 'int'>
```
So easy enough now to check 
```
>>> mpy[0]==77
True
```
Then I wondered if it was possible to check by reference to the 'M' representation? Turns out it is
```
>>> 'M' in mpy[:1]
True
```
I have no idea why it's necessary to reference the 1st 2 bytes to see if the 1st byte is correct using the 'M' alias but I'd be keen to learn if someone would care to explain it?

**Comment by mattytrentini** [comment]
Looks like a minor bug: that comparison is not allowed in Python (tested with v3.5 and v3.8):

```python
>>> 'M' in mpy[:1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: a bytes-like object is required, not 'str'
```

**Comment by karfas** [comment]
Just for the record: `'M' in mpy[:1]` works also in python 2.7.

In response to the title, **who cares about that**, especially in this case ? 

If the file signature is `b'M\x05'`, the simple check for the signature is `mpy == buffer[:2]` and I see no need to fiddle with the single bytes in `mpy`.
If the file signature is only the `'M'`, `mpy = b'M\x05'` is useless anyway; the comparison comes down to `ord('M') == buffer[0]` or (less easy to comprehend) `77 == buffer[0]`.

**Comment by jimmo** [comment]
> I have no idea why it's necessary to reference the 1st 2 bytes to see if the 1st byte is correct using the 'M' alias but I'd be keen to learn if someone would care to explain it?

@kjm1102 Just to clarify something here that might help -- when you write `mpy[:1]` (i.e. slicing) you are referencing only the _first_ byte. (The end is exclusive). `mpy[:x]` can be read as "make a new `bytes` containing up to the start but not including of the x'th byte (where x is zero-indexed of course)".

What @karfas said is correct though -- if you want to compare _byte values_ then you should access individual bytes and compare them by their actual values. i.e. `mpy[0]==77` (or `ord('M')` if you don't want to magic number the 77).

There are a few places in MicroPython where, for code size minimisation, we allow strings and bytes to be interchangeable. i.e. CPython doesn't allow you to mix bytes and strings with the `in` operator as @mattytrentini pointed out (but MicroPython, does specifically only when the str is the LHS and the bytes is the RHS).

The main thing to remember is in Python, there is no "character" type. This is a deliberate decision which is a lot more sensible when encoding is involved. So when you _index_ a string you get a string, and when you index a bytes/bytearray/memoryview you get an int. When you _slice_ either, you get the original type.
